### ISSUE TITLE
Flexible Breakpoint config

### ISSUE THREAD

**Issue by src-code** [yahoo/atomizer#71]
@thierryk recently pointed out that Atomizer's breakpoint design, which limits you to three breakpoints, is not flexible enough to accommodate common use cases.  We probably should allow developers to create unlimited arbitrary breakpoints.  

Also, it'd be a good opportunity to revisit whether it makes sense to have any default breakpoints (such as the 3 we have now), and if so, what their default values should be.

@renatoi

**Comment by renatoi** [comment]
Breakpoints idea in Atomizer was inspired by Bootstrap Grid System:
http://getbootstrap.com/css/#grid-options

Allowing ultra flexible breakPoints might seem a good idea but I wonder if that's ideal/good practice.

There are some problems:

1. Conflicting Breakpoints: If two components declare two different breakpoints with the same name but with different values, which one should be used? I think it makes more sense if the app responsible to lay out the grid is actually the solely owner of breakpoint declarations.

2. If a component declares its own unique breakpoints, then we have bloat. The more we go towards custom rules the more bloat we have. On the other hand, it sucks to limit how the component react to different viewport sizes (Ideally, it should react to its OWN size, not the viewport, but this is not doable with CSS alone at the moment and very costly to do it with JS). So we either limit the capability of breakpoints on purpose to enforce good practice or we allow custom breakpoints to be declared at the cost of bloat. Keep in mind that the bloat I'm talking about here is not only in the CSS, but also in the markup.

I honestly prefer to keep having 3 breakpoints defined by the top level component and if any child component needs to have its own set of custom rules, then they should create their own custom CSS (not atomic). At least the bloat is only in one place.

**Comment by thierryk** [comment]
I think for now we should leave things they way they are and see if users open issues about this and what are their asks exactly. 

As a side note, I'm not really in favor of *enforcing* good practice. In my opinion, best practice should be a *suggestion*, but not be imposed onto the user. After all, there are exceptions to every rules...

**Comment by renatoi** [comment]
@thierryk suggested creating an option for the breakpoints so the rules are contained in the media query.

```css
/**
 * computed if:
 * - no responsive class have been used OR
 * - it is not computed due to MQ conditions.
**/
.D-b {
    display: block;
}

/* computed once viewport is at least 300px */
/* if we grow the viewport, this rule will still be computed one to the element it is applied */
@media (min-width: 300px) {
    .D-b--sm {
        display: block;
    }
}

/* computed only if the viewport width is less than 300px */
/* if we grow the viewport, this rule won't be computed once we get past 300px */
@media (max-width: 300px) {
    .D-b--sm {
        display: block;
    }
}

/* computed only if the viewport width is between 100px and 300px */
@media (min-width: 100px and max-width: 300px) {
    .D-b--sm {
        display: block;
    }
}
```

For this, I propose the following example:

```js
{
    breakpoints: {
        sm: '@media(min-width: 100px)',
        md: '@media(min-width: 100px and max-width: 400px)',
        lg: '@media(max-width: 500px)'
    }
}
```

Does it solve the problem you had @thierryk?

**Comment by renatoi** [comment]
Ok, so we could also allow custom breakpoints, even though we should say it's a good practice to let the root level component define those as they are mainly the ones who will dictate the layout of the page.

**Comment by thierryk** [comment]
I like that. 
I think it'll solve the problem for most people. The issue we have on our end is that Design often requires *a lot* of "breakpoint granularity" so in such case I think the task should be done via a *layout sheet* rather than via Atomic classes.

**Comment by renatoi** [comment]
Also, they should only be defined in the config, not in the markup, so no things like:

```html
<!-- never like this --->
<div class="D-b--sm(min:100px, max: 400px)"></div>
<!-- always like this -->
<div class="D-b--sm"></div>
```

If `D-b--sm` needs to be custom, configure that in the config.

**Comment by renatoi** [comment]
@thierryk yes, as I said, we should make clear that although it's possible to create custom breakpoints with atomizer, ideally you would want to do that in a custom stylesheet and go non-atomic at that point, so you don't have lengthy markup AND css. A good practice is to let the root level component dictate the layout of the page with the conventional breakpoints (`sm`, `md` and `lg`) whenever possible.

**Comment by src-code** [comment]
Addressed in pull #108

**Comment by chinchang** [comment]
I love this syntax! 😍
